suite/tests/security-common/selfmod.c

```diff
@@ -230,22 +276,39 @@ test_sandbox_cross_page(void)
 static void
 test_sandbox_last_byte(void)
 {
-    int r;
-    byte *last_byte = (byte *)last_byte_jmp_no_ilt + 1;
-    if (!ALIGNED(last_byte, PAGE_SIZE)) {
-        print("laste_byte is not page-aligned: " PFX "\n"
+#    ifdef X86
+    byte *first_instr_of_page = (byte *)last_byte_jmp_no_ilt + 1;
+#    else
+    byte *first_instr_of_page = (byte *)last_byte_jmp_label;
+#    endif
+    if (!ALIGNED(first_instr_of_page, PAGE_SIZE)) {
+        print("first_instr_of_page is not page-aligned: " PFX "\n"
               "Instruction sizes in sandbox_last_byte must be wrong.\n",
-              last_byte);
+              first_instr_of_page);
     }
     print("start last byte test\n");
-    protect_mem(last_byte, PAGE_SIZE, ALLOW_READ | ALLOW_WRITE | ALLOW_EXEC);
+    protect_mem(first_instr_of_page, PAGE_SIZE, ALLOW_READ | ALLOW_WRITE | ALLOW_EXEC);
     /* Execute self-modifying code to create a sandboxed page. */
+#    ifdef X86
     make_last_byte_selfmod();
-    r = sandbox_last_byte();
+#    else
+    make_code_page_selfmod();
+#    endif
+
+    int r = sandbox_last_byte();
     print("sandbox_last_byte: %d\n", r); /* Should be 0. */
 
+#    ifdef X86
     /* Make the relative jmp offset zero, so it goes to the next instruction. */
-    *last_byte = 0;
+    *first_instr_of_page = 0;
+#    else
+    /* Increment the branch value, so it jumps to "last_byte_ret_one" */
+    uint32 *first_instruction = (uint32 *)last_byte_jmp_label;
+    (*first_instruction)++; /* The jump distance is encoded in number of instructions, not
+                               bytes. */
+    tools_clear_icache(first_instruction, first_instruction + 4);
```

Comment:
I don't think self modifying code is expected to work on AARCH64 without clearing the icache? I'll add a comment.

Reply:
AArch64 processors are allowed to keep the icache consistent in hardware, in which case the software does *not* need to explicitly clear the icache.  This is what the issue you saw that I linked to, #5771, covers.  Supporting such a processor is where we'd end up needing DR"s page prot and sandboxing features.

Reply:
I have added a wrapper function around tools_clear_icache() so that in the future we can avoid calling tools_clear_icache if it is not architecturally necessary.